### DISCUSSION TITLE
Testing fireAndForget effect

### DISCUSSION THREAD

**Discussion by maxkazakov** [pointfreeco/swift-composable-architecture#1308] [Q&A] [answered]
Hey! I use version 0.39.1 with the Effects on Publishers.

I have a reducer which is responsible for saviging data to file:
```
let workoutPlansListSaveReducer = Reducer<WorkoutPlansList, WorkoutPlansListAction, WorkoutPlansListEnvironment> {  state, action, env in
    enum SaveToken {}
    return env.workoutPlansStorage.store(state.workoutPlans.elements)
        .debounce(id: SaveToken.self, for: 0.5, scheduler: env.mainQueue)
        .fireAndForget()
}
```

 This reducer reacts on every action and it is combined with main logic reducer.
 
I want to test only main logic reducer without taking into account workoutPlansListSaveReducer's Effect, and I obviously receive "An effect returned for this action is still running." error. So, I have to append `self.mainQueue.advance(by: .0.5)` to the end of every test case.

Is there a smarter way to make test passed (without just advancing scheduler) ?
Is there a way to test workoutPlansListSaveReducer's Effect?

**Comment by Alex293** [comment]
Just pass you main reducer to the test store instead of passing the merged reducers

**Reply by maxkazakov** [reply]
Thanks! This is one of the option. But I'm still looking for a way to test combined reducer and to make sure that fireAndForget effect was called.

**Comment by mbrandonw** [comment, marked as answer]
Have you tried using an [`immediate`](https://github.com/pointfreeco/combine-schedulers/blob/main/Sources/CombineSchedulers/ImmediateScheduler.swift) scheduler for there `mainQueue` during tests? For example, this test passes:

```swift
func testImmediateDebounce() async {
  enum Action { case tap }
  let store = TestStore(
    initialState: 0,
    reducer: Reducer<Int, Action, Void> { state, action, _ in
      return .fireAndForget {
      }
      .debounce(id: 1, for: 1, scheduler: DispatchQueue.immediate)

    },
    environment: ()
  )

  await store.send(.tap)
}
```

**Reply by maxkazakov** [reply]
This test passes even if I just return .none from the reducer. Is there a way to make sure that .fireAndForget effect was invoked?

**Reply by mbrandonw** [reply]
Oh yes, you would mock the `env.workoutPlansStorage.store` endpoint to mutate a bit of local state in your test, and then assert on it after.

In my example above this would look roughly like the following:

```swift
func testImmediateDebounce() async {
  enum Action { case tap }
  let store = TestStore(
    initialState: 0,
    reducer: Reducer<Int, Action, Environment> { state, action, environment in
      return .fireAndForget {
        environment.save()
      }
      .debounce(id: 1, for: 1, scheduler: DispatchQueue.immediate)

    },
    environment: ()
  )

  var saveCount = 0
  store.environment.save = { _ in saveCount += 1 }

  await store.send(.tap)
  XCTAssertEqual(saveCount, 1)
}
```

**Reply by maxkazakov** [reply]
Thank you!
